crates/rpc/rpc/src/eth/api/mod.rs

```diff
@@ -448,7 +466,251 @@ struct EthApiInner<Provider, Pool, Network> {
     pending_block: Mutex<Option<PendingBlock>>,
     /// A pool dedicated to blocking tasks.
     blocking_task_pool: BlockingTaskPool,
+    /// Cache for block fees history
+    fee_history_cache: FeeHistoryCache,
     /// An http client for communicating with sequencers.
     #[cfg(feature = "optimism")]
     http_client: reqwest::Client,
 }
+
+/// Settings for the [FeeHistoryCache](crate::eth::FeeHistoryCache).
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FeeHistoryCacheConfig {
+    /// Max number of blocks in cache.
+    ///
+    /// Default is 1024.
+    pub max_blocks: u64,
+}
+
+impl Default for FeeHistoryCacheConfig {
+    fn default() -> Self {
+        FeeHistoryCacheConfig { max_blocks: 1024 }
+    }
+}
+
+/// Wrapper struct for BTreeMap
+#[derive(Debug, Clone)]
+pub struct FeeHistoryCache {
+    lower_bound: Arc<AtomicU64>,
+    upper_bound: Arc<AtomicU64>,
+    config: FeeHistoryCacheConfig,
+    entries: Arc<tokio::sync::RwLock<BTreeMap<u64, FeeHistoryEntry>>>,
+    eth_cache: EthStateCache,
+}
+
+impl FeeHistoryCache {
+    /// Creates new FeeHistoryCache instance, initialize it with the mose recent data, set bounds
+    pub fn new(eth_cache: EthStateCache, config: FeeHistoryCacheConfig) -> Self {
+        let init_tree_map = BTreeMap::new();
+
+        let entries = Arc::new(tokio::sync::RwLock::new(init_tree_map));
+
+        let upper_bound = Arc::new(AtomicU64::new(0));
+        let lower_bound = Arc::new(AtomicU64::new(0));
+
+        FeeHistoryCache { config, entries, upper_bound, lower_bound, eth_cache }
+    }
+
+    /// Processing of the arriving blocks
+    pub async fn on_new_block<'a, I>(&self, blocks: I)
+    where
+        I: Iterator<Item = &'a Block>,
+    {
+        let mut entries = self.entries.write().await;
+
+        for block in blocks {
+            let mut fee_history_entry = FeeHistoryEntry::new(block, block.clone().seal_slow().hash);
+
+            let percentiles = predefined_percentiles();
+
+            fee_history_entry.rewards = calculate_reward_percentiles_for_block(
+                &percentiles,
+                &fee_history_entry,
+                self.eth_cache.clone(),
+            )
+            .await
+            .unwrap_or_default();
```

Comment:
this is inefficient for new blocks because this will now fetch the txs and receipts, that we already have from the cache again.

consider rewriting `calculate_reward_percentiles_for_block` so that it takes transactions and receipts